enhancements/20240709-longhorn-v2-data-engine-v1.4.0.md

```diff
@@ -0,0 +1,174 @@
+- We actually don't want to always create `/oem/93_spdk.yaml` on every node, because this will allocate 1024 huge pages (i.e. it will eat 2GiB RAM), even if the V2 Data Engine is not enabled.
+    - We should be able to solve this by teaching node-manager to create/delete `/oem/93_spdk.yaml`, depending on whether the V2 Data Engine is enabled, and whether or not a given node has the `node.longhorn.io/disable-v2-data-engine` label set to `true`.
+
+To implement all the above, we can:
+- In harvester:
+    - Register a syncer for the `longhorn-v2-data-engine` setting. When the value of the setting changes, we do two things:
+        1. Update the NodeConfig for each node to indicate that the V2 Data Engine should be enabled/disabled, taking into account the value of the node's `node.longhorn.io/disable-v2-data-engine` label.  This change will then be picked up by node-manager, much as is currently done for NTP config.
+        2. Write the setting to Longhorn's `lhs/v2-data-engine` setting.
+    - Watch for changes to node labels.  If `node.longhorn.io/disable-v2-data-engine` changes, again, update the NodeConfig for each node appropriately.
+- In node-manager:
+    - When a given NodeConfig changes to enable the V2 engine:
+        - Create `/oem/93_spdk.yaml`
+        - Run `modprobe vfio_pci ; modprobe uio_pci_generic ; modprobe nvme_tcp`
+        - Run `echo 1024 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+        - If any of the modprobes fail, or the hugepages can't be allocated, record this somewhere so that it can be picked up in the Harvester GUI with a suggestion to the user to reboot the node (TODO: figure out how to actually do this).
+    - If the V2 engine becomes disabled:
+        - Remove `/oem/93_spdk.yaml`
+        - Run `modprobe -r vfio_pci ; modprobe -r uio_pci_generic ; modprobe -r nvme_tcp`
+        - Run `echo 0 > /sys/kernel/mm/hugepages/hugepages-2048kB/nr_hugepages`
+
+[os2#98]: https://github.com/harvester/os2/pull/98
+[Longhorn Quick Start]: https://longhorn.io/docs/1.6.2/v2-data-engine/quick-start/
+
+#### Adding Disks ####
+
+This is largely covered by Vicente's [node-disk-manager refactor HEP], which proposes adding a Longhorn V2 provisioner to NDM.
+
+Note that we will also need to teach NDM to ignore V2 volumes attached to the harvester host, much as it currently ignores V1 volumes by looking for the string "longhorn" in the block device's bus path.  For V2 volumes, we end up with a bunch of `/dev/nvme*` and `/dev/dm-*` devices.  We should probably just ignore the latter by device path.  For nvme devices, here's an example of the udev data, which suggests we can look for "SPDK_bdev_Controller" in a variety of fields:
```

Comment:
To filter out the v1/v2 volume, we could leverage this issue: https://github.com/harvester/harvester/issues/5059.
Using the latest filter mechanism to filter out the v1/v2 volume.